tapsend/allocation.go

```diff
@@ -389,27 +391,29 @@ func sortPiecesWithProofs(pieces []*piece) {
 	// Now sort all the proofs within each piece by amount and then script
 	// key. This will give us a stable order for all asset UTXOs.
 	for idx := range pieces {
-		sort.Slice(pieces[idx].proofs, func(i, j int) bool {
-			assetI := pieces[idx].proofs[i].Asset
-			assetJ := pieces[idx].proofs[j].Asset
-
-			// If amounts are equal, sort by script key.
-			if assetI.Amount == assetJ.Amount {
-				keyI := assetI.ScriptKey.PubKey
-				keyJ := assetJ.ScriptKey.PubKey
-				return bytes.Compare(
-					keyI.SerializeCompressed(),
-					keyJ.SerializeCompressed(),
-				) < 0
-			}
-
-			// Otherwise, sort by amount, but in reverse order so
-			// that the largest amounts are first.
-			return assetI.Amount > assetJ.Amount
-		})
+		slices.SortFunc(
+			pieces[idx].proofs, func(i, j *proof.Proof) int {
+				return InputSort(i.Asset, j.Asset)
+			},
+		)
 	}
 }
 
+// InputSort is a comparison function used to sort asset inputs by amount and
+// then by script key. This is used to ensure that the inputs are sorted in a
+// deterministic way before they are distributed to the outputs.
+func InputSort(i, j asset.Asset) int {
+	return cmp.Or(
+		// Sort amounts in reverse order so that the largest amounts are
+		// first.
+		cmp.Compare(j.Amount, i.Amount),
+		bytes.Compare(
+			i.ScriptKey.PubKey.SerializeCompressed(),
+			j.ScriptKey.PubKey.SerializeCompressed(),
+		),
+	)
+}
+
```

Comment:
You could generalise the function name and docstring to `AssetSort`, as it just sorts assets. IMO since it's a pure function, there's no need to explain its usage in the doc.

Reply:
The way we sort the assets is kind of specific to inputs. And the idea is that we should use this same function everywhere we deal with inputs, that why I wanted to have the word "Input" in there. I renamed to `AssetSortForInputs`, hope that makes things more clear.